src/mlpack/tests/main_tests/fastmks_test.cpp

```diff
@@ -0,0 +1,658 @@
+  SetInputParam("k", (int) 10);
+  SetInputParam("single", true);
+
+  mlpackMain();
+
+  CheckMatrices(indices,
+      CLI::GetParam<arma::Mat<size_t>>("indices"));
+  CheckMatrices(kernel,
+      CLI::GetParam<arma::mat>("kernels"));
+}
+
+/*
+ * Ensure that we get almost same results in cover tree
+ * search mode when different basis is specified.
+ */
+BOOST_AUTO_TEST_CASE(FastMKSBasisTest)
+{
+  // 100 points in 3 dimensions.
+  arma::mat referenceData(3, 100, arma::fill::randu);
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", referenceData);
+  SetInputParam("k", (int) 10);
+  SetInputParam("base", 3.0);
+
+  mlpackMain();
+
+  arma::Mat<size_t> indices;
+  arma::mat kernel;
+  indices = std::move(CLI::GetParam<arma::Mat<size_t>>("indices"));
+  kernel = std::move(CLI::GetParam<arma::mat>("kernels"));
+
+  bindings::tests::CleanMemory();
+
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["k"].wasPassed = false;
+
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("base", 4.0);
+
+  mlpackMain();
+
+  arma::Mat<size_t> newindices;
+  arma::mat newkernel;
+  newindices = std::move(CLI::GetParam<arma::Mat<size_t>>("indices"));
+  newkernel = std::move(CLI::GetParam<arma::mat>("kernels"));
+
+  CheckMatrices(indices, newindices);
+  CheckMatrices(kernel, newkernel);
+}
+
+/**
+  * Check that we can't specify base less than 2.
+  */
+BOOST_AUTO_TEST_CASE(FastMKSBaseTest)
+{
+  // 100 points in 3 dimensions.
+  arma::mat referenceData(3, 100, arma::fill::randu);
+
+  // Random input, invalid base.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("base", 0.0); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::invalid_argument);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+  * Ensure that different kernels returns different results.
+  */
+BOOST_AUTO_TEST_CASE(FastMKSKernelTest)
+{
+  std::string kerneltypes[] = {"polynomial", "cosine", "gaussian",
+      "epanechnikov", "triangular", "hyptan"};
+  const int nofkerneltypes = 6;
+  // 100 points in 3 dimensions.
+  arma::mat referenceData(3, 100, arma::fill::randu);
+  // 90 points in 3 dimensions.
+  arma::mat queryData(3, 90, arma::fill::randu);
+  // Keep some k <= number of reference points same over all.
+  SetInputParam("k", (int) 10);
+
+  arma::Mat<size_t> indices_compare;
+  arma::mat kernels_compare;
+
+  arma::Mat<size_t> indices;
+  arma::mat kernels;
+
+  // Looping over all the kernels
+  for (int i = 0; i < nofkerneltypes; i++)
+  {
+    // Same random inputs, different algorithms.
+    SetInputParam("reference", referenceData);
+    SetInputParam("query", queryData);
+    SetInputParam("kernel", kerneltypes[i]);
+
+    mlpackMain();
+
+    if (i == 0)
+    {
+      indices_compare =
```

Comment:
Tiny style issue: please don't use variables with `_` in the name; instead call this `indicesCompare` or similar. :+1: